### PR TITLE
Fixes for compiling with a newer (system) jsoncpp

### PR DESCRIPTION
**Pull request by Rogier-5**
See issue #4306.

The primary problem turns out to be that minetest uses `#include "json/json.h"`, where it should use `#include <json/json.h>`.

Another problem which surfaces then, is that jsoncpp contains a bug, which makes it fail to compile in non-c++11 mode. The following change is neede to `/usr/include/jsoncpp/json/config.h`, to make it compile. 

Change:
```C++
#if defined(_MSC_VER) && _MSC_VER <= 1600 // MSVC <= 2010
# define JSONCPP_OVERRIDE
#else
# define JSONCPP_OVERRIDE override
#endif // MSVC <= 2010
```
To:
```C++
#if __cplusplus < 201103L
# define JSONCPP_OVERRIDE
#else
# define JSONCPP_OVERRIDE override
#endif // __cplusplus < 201103L
```

Jsoncpp's config.h contains other similar code (similar to `#if defined(_MSC_VER) && _MSC_VER <= 1600 // MSVC <= 2010`), and I haven't assessed whether that should be changed as well. With the jsoncpp patch above, minetest compiles, and the server starts and seems to work fine. I couldn't test the client.

### CODE REVIEW

File: src/script/common/c_content.cpp [hunk 1]
```diff
@@ -1250,8 +1250,13 @@ static bool push_json_value_helper(lua_State *L, const Json::Value &value,
 			lua_newtable(L);
 			for (Json::Value::const_iterator it = value.begin();
 					it != value.end(); ++it) {
+				#ifndef JSONCPP_STRING
```
Comment (est31) [inline src/script/common/c_content.cpp]: Can you put these entirely to the right?

**Comment by est31** [comment]
:+1:

**Comment by nerzhul** [comment]
:+1:

**Comment by zing269** [comment]
Using Arch with gcc 5, I was unable to compile for android, many instances of unable to find json/json.h. I manually backed out the subdirectory changes and was then able to compile successfully.

**Comment by Rogier-5** [comment]
@zing269 This sounds like you compiled with `ENABLE_SYSTEM_JSONCPP` set to true (or 1), but your system does not have a suitable version of jsoncpp installed.

Please try `cmake -DENABLE_SYSTEM_JSONCPP=0`.

**Comment by zing269** [comment]
@Rogier-5 I tried your suggestion but I'm still getting the No such file or directory error. I found a couple of entries in /build/android/jni/Android.mk that referenced /jni/src/json and modified those to /jni/src/jsoncpp/json but that didn't help either.

**Comment by Rogier-5** [comment]
> modified those to /jni/src/jsoncpp/json but that didn't help either.

I think that for `LOCAL_SRC_FILES`, you need
```C++
LOCAL_SRC_FILES += jni/src/jsoncpp/json/jsoncpp.cpp
```
(which how you probably changed it already)

and for `LOCAL_C_INCLUDES`, you need:
```C++
[...]
jni/src/jsoncpp                           \
[...]
```
which you probably changed to `jni/src/jsoncpp/json` instead ?

Does that help ?

**Comment by zing269** [comment]
Changing the LOCAL_C_INCLUDES to jni/src/jsoncpp did the trick. Thanks.

**Comment by est31** [comment]
Mhhh these things need to be done on master as well...

**Comment by Rogier-5** [comment]
@est31 naturally

@zing269 Are you planning to create a pull request ?

If not: I pushed a fix, but I haven't been able to compile-test it yet. Would you mind verifying that it works for you ? It is at https://github.com/Rogier-5/minetest/tree/json-android-fix.

**Comment by est31** [comment]
@Rogier-5 I confirm the android build works with your patch.

**Comment by zing269** [comment]
@Rogier-5 Please go ahead and create the pull request. I'm still figuring out this whole git thing.

**Comment by Rogier-5** [comment]
In the mean time, I have managed to configure the android environment as well, and I con confirm too that compilation fails without the patch, and succeeds with patch.

@zing269: the following shell commands should work:
```
git clone https://github.com/minetest/minetest
cd minetest
git fetch https://github.com/Rogier-5/minetest json-android-fix:json-android-fix
git checkout json-android-fix
cd build/android
make
```
(if that's of any use)

**Comment by zing269** [comment]
@Rogier-5 thanks, that is quite helpful.

**Pull request merged**